### PR TITLE
[blog] How to Deprecated method - open source

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

**Comment by TomasVotruba** [comment]
@enumag Could I ask for a review? I think you might know about this one

File: source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+```
+
+Notice `@`!
+
+So the error is there for the user, but it is suppressed? WTF!
+
+
+### Step by Step
+
+It's related to release process. Just imagine this life cycle of `add()` method:
```
Comment (enumag) [inline source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md]: I don't think Symfony uses the at-sign to achieve this. And I don't like such gradual deprecations by the way.

The standard `trigger_error()` call means that the notices are opt-out. You need to disable them with `error_reporting`. Symfony uses at-signs to make them opt-in instead - you will only see them if you use an error handler that would make them visible. See https://symfony.com/doc/current/components/phpunit_bridge.html#trigger-deprecation-notices.

The reason why I don't like this Step by Step approach is that you would then mix both opt-in and opt-out deprecation notices in your code which seems like a very bad practice. Please choose just one and be consistent about it. I would recommend removing this Step by Step section and rewrite it with opt-in vs opt-out behaviour, perhaps with an example of the opt-in error handler.

It's just a metter of opinion which one you prefer. I personally prefer the opt-in approach.

File: source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+---
+
+This technique is quite rare to see (apart PHP frameworks). It's very simple to add to your open-source code workflow though - let me convince you.
+
+
+## Why Write Deprecation Messages?
+
+## If you don't
+
+- people will have to find themselves in commits - from programmer to detective
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md]: Well that's 80% of my upgrade experience :D pain fire:

File: source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+```php
+/**
+ * @deprecated Method add() is deprecated, use addHtml() instead.
+ */
+public function add(...)
+```
+
+
+**Nice!** Even I can understand this now and I'm very poor English.
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md]: > Maybe shorten it by one step

Allright! Thanks for this

File: source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md (hunk 1 above)
Comment (TomasVotruba) [reply source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md]: > I don't think Symfony uses the at-sign to achieve this. And I don't like such gradual deprecations by the way.

Oh, thanks for the link!

> The reason why I don't like this Step by Step approach is that you would then mix both opt-in and opt-out deprecation notices in your code which seems like a very bad practice.

You're right. It's bit confusing, now when I read Symfony code I see them both with no logic behind that.

>  I would recommend removing this Step by Step section and replace it with opt-in vs opt-out section, perhaps with an example of the opt-in error handler.

I'll probably just remove it. This is more complicated then I thought and personally I use them without `@` everywhere.

File: source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md [hunk 4]
```diff
@@ -212,92 +212,8 @@ I said...
...
-}
-```
-
-**4.0**
-
-```php
-public function addHtml(...)
-{
-}
-```
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md]: I see the cyclomatic complexity of this post would be like 10x higher with this section.

Thanks @enumag for prescious feedback! Feels much better to read now

**Comment by enumag** [comment]
Yeah you're right the opt-in / opt-out section would raise the complexity of this article. Still though as it is the article is quite useless for experienced developers where as the `@` tricks are not so widely known in my opinion. Do you think it could be long enough for a separate article?

**Comment by TomasVotruba** [comment]
> Still though as it is the article is quite useless for experienced developers...

It aims on people who don't make their own frameworks or packages yet. Personally I've seen it only in Symfony and Nette. 

> ... the @ tricks are not so widely known in my opinion. Do you think it could be long enough for a separate article?

Advanced posts for the next-level group would be great though. I don't feel like the right person to write about it, since I didn't know till you told me, but I think all has been already written in my post + your comments.

I think your experince should be shared. Would you take it and put a post together on Pehapkari blog? That would be :party

**Comment by enumag** [comment]
Oki. I think I could find some time for this since the article probably won't be too long.

**Comment by TomasVotruba** [comment]
:clap: 

Feel free to reuse any parts of the "bonus section": https://github.com/TomasVotruba/tomasvotruba.cz/pull/154/commits/746cb3f689be8bbdd3a33d772a299c78e809d1d4

**Pull request merged**